sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -69,8 +80,29 @@ public bool IsRunning
                     return _isRunningOverride.Value;
                 }
 
-                // TODO: Implement
-                throw new NotImplementedException();
+                // If there is no active processor task, ensure that it is not
+                // in the process of starting by attempting to acquire the semaphore.
+                //
+                // If the semaphore could not be acquired, then there is an active start/stop
+                // operation in progress indicating that the processor is not yet running or
+                // will not be running.
+
+                if (_runningProcessorTask == null)
+                {
+                    try
+                    {
+                        if (!ProcessorRunningSync.Wait(100))
+                        {
+                            return false;
+                        }
+                    }
+                    finally
+                    {
+                        ProcessorRunningSync.Release();
+                    }
+                }
+
+                return  ((!_runningProcessorTask?.IsCompleted) ?? (false));
```

Comment:
I've toyed with the idea and haven't found a variant that I like.   I'm certainly open to it, and am going to take this opportunity to bounce some thoughts around and pick your brain.   If we end up in a place that we think is solid, I'll try to champion it for approval.  

It feels like using the actual `TaskStatus` is both leaking some implementation details and presenting a potentially confusing list with more states that we really would want.   The status values that would make sense to me are a set like:  `Running`, `Faulted`, `NotRunning` which just offers a bit of clarity for why things may not be running.    Do you see any others that would potentially be useful for the processor itself (pretending that we don't know the underlying implementation is a background task) ?

We intentionally don't try to expose concepts like how many partitions are owned and such, instead asking developers to track what is interesting to them via context from the processor's handlers.    In the case of "my processor just entered a faulted state" the minimal would be:

  - See that the error handler gets called without a partition context
  - Check `IsRunning` from the error handler to see it was fatal.

You could get a bit more insight by:

 - See that the initialize gets called for some partitions
 - See that the error handler gets called without a partition context
 - See that every partition that was initialized was released
 - Check `IsRunning` from the error handler or in response to the unexpected mass release of partitions

Since we definitely encourage making decisions in the error handler around the state of the processor, and have concerns that folks will throw an empty implementation there and ignore things.  I would say the minimal case is a common thing and something we'd want to encourage regardless of `Status` over `IsRunning`.    Since that kind of obviates the `Faulted` status, more-or-less, we're back to `Running` and `NotRunning` as the core bits of information that we'd be conveying.

What I'd ask is if there are scenarios that you're thinking of where additional information would make sense?   Would those scenarios lessen the emphasis on the error handler as a key focus?

Reply:
> It feels like using the actual `TaskStatus` is both leaking some implementation details and presenting a potentially confusing list with more states that we really would want. The status values that would make sense to me are a set like: `Running`, `Faulted`, `NotRunning` which just offers a bit of clarity for why things may not be running. Do you see any others that would potentially be useful for the processor itself (pretending that we don't know the underlying implementation is a background task) ?

I agree totally about not leaking the task status directly. Below are some statuses that I was thinking could eliminate some ambiguity around the state of the processor without leaking implementation:
- `NotStarted`
  - Processor initialized but Start has not been called
- `Starting`
  - Start called, but not complete, which was the root of the recent timing issue discovered)
- `Running`
  - Same as `IsRunning` == true
- `Stopping`
  - Stop called but not complete (which currently is potentially another race between task state and the stop completion)
- `Stopped`
  - Same as `IsRunning` == false

I hadn't considered leveraging `Status` for any scenarios around faults, but the context you provided there makes sense to me.

Reply:
This is an interesting idea, and I like it.   I hadn't considered using it for transitions in the way that you outlined.  The discussion now gets to turn to how we should expose it.   The big challenge that I see is that we have a processor client derivative with `IsRunning` which we can't take away easily.

The options that I see: 

- Internal is easy;  we just add it.  It's helpful for testing/debugging but provides no value to developers building a processor.

- Protected would require approval, but should be an easy conversation.  This gives developers building a processor insight, but is not surfaced to users of the processor without a developer explicitly choosing to do so.  It may not provide a ton of value in normal use this way, however.   Do we think we have scenarios to justify it?

- Public, alongside the `IsRunning` property for the `EventProcessorClient`.  I don't like this, as we're now exposing two different variations on a similar topic and a busier API surface.  This would potentially be difficult to get architect approval on without some very strong scenarios about why the average processor user would need it.

- Public, removing the `IsRunning` property.  As a breaking change, to the `EventProcessorClient` I think the reasoning against this is obvious.  It would be a very hard sell to get this approved.


I'd very much appreciate your thoughts.

Reply:
Thanks for laying out the context and tradeoffs as you see them. 

Maybe these are scenarios that are mostly relevant for tests and debugging, as opposed to real world developer scenarios. Would mapping the public state of `IsRunning` to these more granular internal statuses have provided more clarity when reasoning about why `IsRunning` is true or false? In tracking down the recent nightly build test issues, I think it would have. I also just could be over thinking it 😄.

Reply:
Couldn't hurt to try it out.   I'll see if I can take the 'internal' route with the next set of changes and see where we end up.